modules/create_timepoint/php/ajaxtimepoint.class.inc

```diff
@@ -0,0 +1,284 @@
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        /**
+         * Ensure it's a POST request.
+         */
+        if ($request->getMethod() != 'POST') {
+            return (new \LORIS\Http\Response())
+                ->withHeader('Content-Type', 'application/json')
+                ->withStatus(405)
+                ->withHeader('Allow', 'POST')
+                ->withBody(
+                    new \LORIS\Http\StringStream(
+                        'Only POST requests are allowed'
+                    )
+                );
+        }
+        /**
+         * Parse POST request body.
+         */
+        $requestBody = $request->getParsedBody();
+
+        /**
+         * User permission verification.
+         */
+        $user = $request->getAttribute('user');
+        if (!isset($requestBody['identifier'])) {
+            return (new \LORIS\Http\Response())
+                ->withHeader('Content-Type', 'application/json')
+                ->withStatus(404)
+                ->withBody(
+                    new \LORIS\Http\StringStream(
+                        'Failed to retrieve data for create timepoint.'
+                    )
+                );
+        }
+        $requestBody['identifier'] = isset($requestBody['identifier']) ?
+            intval($requestBody['identifier']) : null;
+        $requestBody['psc']        = empty($requestBody['psc']) ?
+            null : $requestBody['psc'];
+        $candidate = \Candidate::singleton(
+            $requestBody['identifier']
+        );
+        if (!$user->hasPermission('data_entry')
+            && !(in_array(
+                intval($candidate->getData('RegistrationCenterID')),
+                $user->getData('CenterIDs'),
+                true
+            ))
+        ) {
+            return (new \Loris\Http\Response())
+                ->withStatus(403, 'Forbidden')
+                ->withBody(
+                    new \LORIS\Http\StringStream(
+                        json_encode(
+                            array('error' => 'Forbidden')
+                        )
+                    )
+                );
+        }
+
+        /**
+         * Response to reply.
+         */
+        $response = $this->processRequest($requestBody);
+        return (new \LORIS\Http\Response())
+            ->withHeader('Content-Type', 'application/json')
+            ->withStatus(200)
+            ->withHeader('Allow', 'POST')
+            ->withBody(
+                new \LORIS\Http\StringStream(
+                    json_encode($response)
+                )
+            );
+    }
+
+    /**
+     * Processes the values and saves to database
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function processRequest(array $values) : array
+    {
+        if (isset($values['command']) && $values['command'] == 'initialize') {
+            $response = $this->initializeSetup($values);
```

Comment:
This should not be a POST. If the goal is to get values from the server then GET should be used.